staging/src/k8s.io/apimachinery/pkg/util/wait/loop_test.go

```diff
@@ -203,45 +204,89 @@ func Test_loopConditionUntilContext_semantic(t *testing.T) {
 			attemptsExpected: 0,
 			errExpected:      context.DeadlineExceeded,
 		},
+		{
```

Comment:
From the coverage profile it looks like we're missing coverage in one scenario:

> sliding=true, condition fails one time

Can you add tests for 

1. sliding=true, condition fails once, then passes
2. sliding=true, condition fails once, then has a context cancellation error

Reply:
Ok, with the right version checked out coverage shows:

```
        for {
                // checking ctx.Err() is slightly faster than checking a select
                if err := ctx.Err(); err != nil {
                        return err // uncovered
                }
```

This now immediately follows

```
	select {
	case <-doneCh:
		return ctx.Err()
	case <-timeCh:
	}
```

and is redundant (before, we could conditionally check).  We also check this before the loop.  So the only condition this `if` block can be called in now is when we wait the full duration of `<-timeCh` and then in the nanoseconds between that check and condition the context is cancelled.  That's not necessary, so we should remove the first if block in the main `for {}` loop.

Reply:
> That's not necessary, so we should remove the first if block in the main for {} loop.
Updated.

Reply:
Remove this caused some flaky tests:
```
{Failed;Failed;Failed;  === RUN   Test_loopConditionUntilContext_semantic/context_cancelled_after_5_attempts
    loop_test.go:286: expected attempts count: 5 but got: 6
    --- FAIL: Test_loopConditionUntilContext_semantic/context_cancelled_after_5_attempts (0.00s)
;=== RUN   Test_loopConditionUntilContext_semantic/context_cancelled_and_immediate_after_5_attempts
    loop_test.go:286: expected attempts count: 5 but got: 6
    --- FAIL: Test_loopConditionUntilContext_semantic/context_cancelled_and_immediate_after_5_attempts (0.00s)
;=== RUN   Test_loopConditionUntilContext_semantic
--- FAIL: Test_loopConditionUntilContext_semantic (2.62s)
;}
```
Same as before, I can't reproduce these failures on my mac. I assume it's because the removal of this check. Let me test with the CI.

Reply:
@smarterclayton This removal seems can break some existing uts, so I increased the timer's duration in the tests(see the newest commit).

Reply:
> That's not necessary, so we should remove the first if block in the main `for {}` loop.

@smarterclayton remember that was redundant but was there on purpose, see the comment on line 85. we should keep it

```
		// NOTE: b/c there is no priority selection in golang
		// it is possible for this to race, meaning we could
		// trigger t.C and doneCh, and t.C select falls through.
		// In order to mitigate we re-check doneCh at the beginning
		// of every loop to guarantee at-most one extra execution
		// of condition.
```

Reply:
Hrm.  If I set a really really fast timer, it's a good point that we could never execute.  However, now that we've eliminated the conditional invocation, we should move it up in the loop:

```
	if sliding {
		timeCh = t.C()
	}

	for {
	        // Wait for either the context to be cancelled or the next invocation be called
		select {
		case <-doneCh:
			return ctx.Err()
		case <-timeCh:
		}
		// IMPORTANT: Because there is no channel priority selection in golang
		// it is possible for very short timers to "win" the race in the previous select
		// repeatedly even when the context has been canceled.  We therefore must
		// explicitly check for context cancellation on every loop and exit if true to
		// guarantee that we don't invoke condition more than once after context has
		// been cancelled.
		if err := ctx.Err(); err != nil {
			return err
		}
		
		if !sliding {
			t.Next()
		}
		if ok, err := func() (bool, error) {
			defer runtime.HandleCrash()
			return condition(ctx)
		}(); err != nil || ok {
			return err
		}
		if sliding {
			t.Next()
		}
	}
```

I think this is more correct now (and i reworded the comments to capture that) - please try this out with your new tests.

Reply:
@smarterclayton  Updated, and I think you're right. Now we can pass all the tests without increase the timer's duration.